cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -119,10 +123,18 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
       // Create and run the program
       Program executableProgram = createProgram(cConf, runner, programDescriptor, artifactDetail, tempDir);
       cleanUpTask = createCleanupTask(cleanUpTask, executableProgram);
-      RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(executableProgram, optionsWithPlugins), programId);
+
+      // Publish the program's starting state
+      String twillRunId = options.getArguments().getOption(ProgramOptionConstants.TWILL_RUN_ID);
+      programStateWriter.start(programId.run(runId), options, twillRunId);
+
+      ProgramController controller = runner.run(executableProgram, optionsWithPlugins);
+      RuntimeInfo runtimeInfo = createRuntimeInfo(controller, programId);
       monitorProgram(runtimeInfo, cleanUpTask);
       return runtimeInfo;
     } catch (Exception e) {
+      // Set the program state to an error when an exception is thrown
+      programStateWriter.error(programId.run(runId), new Throwable(e.getMessage()));
```

Comment:
Good point - I wasn't aware that `Exception` was a subclass of `Throwable`. I will use `e` directly.